File: server/controllers/reportController.js
import asyncHandler from 'express-async-handler';
import Report from '../models/Report.js';
import PG from '../models/PG.js';
import { createNotification } from '../utils/notificationHelper.js';

// @desc   Report a PG
// @route  POST /api/reports/:pgId
export const reportPG = asyncHandler(async (req, res) => {
  const { reason } = req.body;
  if (!reason || reason.trim().length < 10) {
    res.status(400);
    throw new Error('Reason must be at least 10 characters');
  }

  const pg = await PG.findById(req.params.pgId);
  if (!pg || pg.isDeleted) {
    res.status(404);
    throw new Error('PG not found');
  }

  const report = await Report.create({
    pg: pg._id,
    reportedBy: req.user._id,
    reason: reason.trim(),
  });

  await createNotification(
    `PG reported: "${pg.title}" — ${reason.substring(0, 60)}`,
    'REPORT_PG',
    req.user._id,
    pg._id
  );

  res.status(201).json({ success: true, message: 'Report submitted', data: report });
});

File: backend/app/controllers/report_controller.py
from app.core.exceptions import AppError
from app.models import collections
from app.services.notification_service import create_notification
from app.utils.bson_utils import serialize_doc, to_object_id
from app.utils.datetime_utils import utc_now


async def report_pg_controller(db, pg_id: str, payload, current_user):
    pg_object_id = to_object_id(pg_id)
    pg = await db[collections.PGS].find_one({"_id": pg_object_id})
    if not pg or pg.get("isDeleted"):
        raise AppError(status_code=404, message="PG not found")

    now = utc_now()
    doc = {
        "pg": pg_object_id,
        "reportedBy": to_object_id(current_user["id"]),
        "reason": payload.reason.strip(),
        "status": "pending",
        "createdAt": now,
        "updatedAt": now,
    }
    result = await db[collections.REPORTS].insert_one(doc)
    doc["_id"] = result.inserted_id

    await create_notification(
        db,
        message=f"PG reported: \"{pg['title']}\" - {payload.reason[:60]}",
        notification_type="REPORT_PG",
        user_id=to_object_id(current_user["id"]),
        related_id=pg_object_id,
    )

    return {"success": True, "message": "Report submitted", "data": serialize_doc(doc)}